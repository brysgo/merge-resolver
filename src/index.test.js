import { mergeResolver } from "./";
import { mergeResolver as mergeResolverDist } from "../dist";
import 'array-flat-polyfill';

describe("build products", () => {
  it("doesn't break", () => {
    const thing1 = {
      __typename: "Query",
      blah: {
        __typename: "Blah",
        bazzez: [
          { __typename: "Baz", a: "b" },
          { __typename: "Baz", a: "c", d: "e" }
        ]
      }
    };
    const thing2 = {
      __typename: "Query",
      blah: {
        __typename: "Blah",
        bazzez: [{ __typename: "Baz", a: "c", d: "f" }]
      }
    };

    const merge = mergeResolverDist({
      Blah: {
        bazzez: (values, join) => {
          const valuesByKey = {};
          values.flat().forEach(val => {
            valuesByKey[val.a] = valuesByKey[val.a]
              ? join(val, valuesByKey[val.a])
              : val;
          });
          return Object.values(valuesByKey);
        }
      },
      Baz: {
        d: values => values.join("")
      },
      typeFromObj: obj => obj && obj.__typename
    });

    expect(merge([thing1, thing2])).toEqual({
      __typename: "Query",
      blah: {
        __typename: "Blah",
        bazzez: [
          { __typename: "Baz", a: "b" },
          { __typename: "Baz", a: "c", d: "fe" }
        ]
      }
    });
  });
});

describe("mergeResolver", () => {
  it("throws an error instead of merging different types", () => {
    const typeA = "A";
    const typeB = "B";
    const merge = mergeResolver({ typeFromObj: o => o });
    expect(() => merge([typeA, typeB])).toThrowErrorMatchingSnapshot();
  });

  it("delegates merge decisions by type", () => {
    const thing1 = {
      __typename: "Query",
      blah: {
        __typename: "Blah",
        bazzez: [
          { __typename: "Baz", a: "b" },
          { __typename: "Baz", a: "c", d: "e" }
        ]
      }
    };
    const thing2 = {
      __typename: "Query",
      blah: {
        __typename: "Blah",
        bazzez: [{ __typename: "Baz", a: "c", d: "f" }]
      }
    };

    const merge = mergeResolver({
      Blah: {
        bazzez: (values, join) => {
          const valuesByKey = {};
          values.flat().forEach(val => {
            valuesByKey[val.a] = valuesByKey[val.a]
              ? join(val, valuesByKey[val.a])
              : val;
          });
          return Object.values(valuesByKey);
        }
      },
      Baz: {
        d: values => values.join("")
      },
      typeFromObj: obj => obj && obj.__typename
    });

    expect(merge([thing1, thing2])).toEqual({
      __typename: "Query",
      blah: {
        __typename: "Blah",
        bazzez: [
          { __typename: "Baz", a: "b" },
          { __typename: "Baz", a: "c", d: "fe" }
        ]
      }
    });
  });
});
