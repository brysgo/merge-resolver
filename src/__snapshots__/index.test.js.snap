// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeResolver supports setting types on arrays 1`] = `
Object {
  "a": Object {
    "one": "two",
  },
  "b": "dope",
}
`;

exports[`mergeResolver throws an error instead of merging different types 1`] = `
"TypeMismatchError: Cannot merge two different types

          Tried to merge \\"B\\" with \\"A\\""
`;
